src/Agent.Plugins/PipelineCache/PipelineCacheTaskPluginBase.cs

```diff
@@ -1,61 +1,135 @@
 using System;
 using System.Collections.Generic;
-using System.IO;
 using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
 using Agent.Sdk;
-using Microsoft.TeamFoundation.Build.WebApi;
 using Microsoft.TeamFoundation.DistributedTask.WebApi;
 using Microsoft.VisualStudio.Services.Agent.Util;
-using Microsoft.VisualStudio.Services.BlobStore.Common;
-using Microsoft.VisualStudio.Services.BlobStore.WebApi;
 using Microsoft.VisualStudio.Services.Common;
-using Microsoft.VisualStudio.Services.Content.Common.Tracing;
-using Microsoft.VisualStudio.Services.WebApi;
+using Microsoft.VisualStudio.Services.PipelineCache.WebApi;
 
 namespace Agent.Plugins.PipelineCache
 {
     public abstract class PipelineCacheTaskPluginBase : IAgentTaskPlugin
     {
+        protected const string RestoreStepRanVariableName = "RESTORE_STEP_RAN";
+        protected const string RestoreStepRanVariableValue = "true";
+
+        private const string SaltVariableName = "AZDEVOPS_PIPELINECACHE_SALT";
+        private const string OldKeyFormatMessage = "'key' format is changing to a single line: https://aka.ms/pipeline-caching-docs";
+
         public Guid Id => PipelineCachePluginConstants.CacheTaskId;
 
         public abstract String Stage { get; }
 
+        internal static (bool isOldFormat, string[] keySegments,IEnumerable<string[]> restoreKeys) ParseIntoSegments(string salt, string key, string restoreKeysBlock)
+        {
+            Func<string,string[]> splitIntoSegments = (s) => {
+                var segments = s.Split(new [] {'|'},StringSplitOptions.RemoveEmptyEntries).Select(segment => segment.Trim());
+                if(!string.IsNullOrWhiteSpace(salt))
+                {
+                    segments = (new [] { $"{SaltVariableName}={salt}"}).Concat(segments);
+                }
+                return segments.ToArray();
+            };
+
+            Func<string,string[]> splitAcrossNewlines = (s) => 
+                s.Replace("\r\n", "\n") //normalize newlines
+                 .Split(new [] {'\n'}, StringSplitOptions.RemoveEmptyEntries)
+                 .Select(line => line.Trim())
+                 .ToArray();
```

Comment:
ParseIntoSegements is being used in the old format and the new format. I thought that the V0 plugin won't support the new format while the v1 plugin will support new and old.

Reply:
I think Will and I landed on the v0 task supporting both (warning on old) and then the v1 task will only support the new format.